File: 666-path-sum-iv/path-sum-iv.py
class Solution:
    def pathSum(self, nums: List[int]) -> int:
        # Sort numbers in descending order to process from leaf nodes to root
        nums.sort(reverse=True)
        
        # List to store paths from leaf to root in reverse order
        listnums = []
        
        # Maximum depth of the tree, derived from the first (largest) element in nums
        maxdepth = nums[0] // 100
        
        # Initialize path sum to accumulate the sum of values for all paths
        pathsum = 0

        # Iterate through each number in the sorted nums
        for num in nums:
            # If the current node is at the maximum depth (leaf node)
            if num // 100 == maxdepth:
                listnums.append([num])  # Start a new path with this leaf node
                pathsum += num % 10     # Add the node's value to the path sum
                continue  # Move to the next node in nums

            # Count to track connections to children in the existing paths
            countnum = 0

            # Traverse the list of paths to check for valid connections from the current node
            for j in range(len(listnums)):
                # Check if the current node is the parent of the last node in listnums[j]
                # Conditions: Same level difference, valid parent-child positioning in the tree
                if ((listnums[j][-1] // 10) % 10 == 2 * ((num // 10) % 10) or
                    (listnums[j][-1] // 10) % 10 == 2 * ((num // 10) % 10) - 1) and \
                   listnums[j][-1] // 100 == num // 100 + 1:
                    
                    listnums[j].append(num)  # Append current node to the path
                    pathsum += num % 10      # Add the node's value to path sum
                    countnum += 1            # Increment count of valid connections
                    
                    # If the current node has connected all valid paths for its level, break
                    if countnum == 2 ** (maxdepth - num // 100):
                        break

            # If no paths were found for the current node, start a new path with it
            if countnum == 0:
                listnums.append([num])
                pathsum += num % 10  # Add value of the current node to path sum

        # Return the accumulated path sum for all paths from root to leaf
        return pathsum
        



        

File: 666-path-sum-iv/path-sum-iv.java
class Solution {
    // Static variables to hold the final sum of path values and the list of paths
    static int finalsum;
    static List<List<Integer>> listnums;

    public int pathSum(int[] nums) {
        // Sort the array to process nodes from leaf to root
        Arrays.sort(nums);
        
        // Determine the maximum depth of the tree from the largest element
        int maxdepth = nums[nums.length - 1] / 100;
        
        // Initialize list of paths and final sum accumulator
        listnums = new ArrayList<>();
        finalsum = 0;

        // Traverse nodes from the last (deepest level) to the first (root level)
        for (int i = nums.length - 1; i > -1; i--) {
            int hundredsi = nums[i] / 100;      // The depth level of the current node
            int tensi = (nums[i] / 10) % 10;    // Position in the current level
            int onesi = nums[i] % 10;           // Value of the current node

            // If the node is at max depth (leaf node), start a new path
            if (hundredsi == maxdepth) {
                List<Integer> newlist = new ArrayList<>();
                newlist.add(nums[i]);
                listnums.add(newlist);          // Add new path with the leaf node
                finalsum += onesi;              // Add node value to final sum
                continue;
            }

            // Count how many paths this node connects to at the next level
            int count = 0;
            for (int j = 0; j < listnums.size(); j++) {
                // Determine the depth and position of the last node in listnums[j]
                int hundredsj = listnums.get(j).get(listnums.get(j).size() - 1) / 100;
                int tensj = (listnums.get(j).get(listnums.get(j).size() - 1) / 10) % 10;

                // Check if the current node is a parent to the last node in the current path
                if ((tensj == 2 * tensi || tensj == 2 * tensi - 1) && hundredsj == hundredsi + 1) {
                    // Add the current node to this path as a parent node
                    List<Integer> tmplist = listnums.get(j);
                    tmplist.add(nums[i]);
                    listnums.set(j, tmplist);  // Update the path with the new node
                    finalsum += onesi;         // Add the value of this node to the final sum
                    count++;
                }

                // Break out if we have added this node to all possible paths for its level
                if (count == Math.pow(2, maxdepth - hundredsi)) {
                    break;
                }
            }

            // If this node did not connect to any paths, start a new path with it
            if (count == 0) {
                List<Integer> newlist = new ArrayList<>();
                newlist.add(nums[i]);
                listnums.add(newlist);
                finalsum += onesi;
            }
        }

        // Return the total sum of all paths
        return finalsum;
    }
}